File: algorithms/algorithms-python/leetcode/Question_0082_Remove_Duplicates_from_Sorted_List_II.py
#!/usr/bin/python
# coding: utf-8

# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head
        node = ListNode(0)
        node.next = head
        pre = node
        while pre.next:
            cur = pre.next
            while cur.next and cur.val == cur.next.val:
                cur = cur.next
            if cur != pre.next:
                pre.next = cur.next
            else:
                pre = pre.next
        return node.next

File: algorithms/algorithms-cplusplus/leetcode/Question_0082_Remove_Duplicates_from_Sorted_List_II.cc

//Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
 public:
  ListNode* deleteDuplicates(ListNode* head) {
    ListNode dummy(0);
    dummy.next = head;
    ListNode *p = &dummy;
    while (p->next) {
      head = p->next;
      while (head->next && head->val == head->next->val) {
        head = head->next;
      }
      if (head == p->next) {
        p = p->next;
      } else {
        p->next = head->next;
      }
    }
    return dummy.next;
  }
};
